ecs-agent/acs/session/task_manifest_responder_test.go

```diff
@@ -0,0 +1,131 @@
+
+func (cid *fakeContainerInstanceDirector) GetLatestSequenceNumber() int64 {
+	return cid.fakeCI.LatestTaskManifestSequenceNumber
+}
+
+func (cid *fakeContainerInstanceDirector) GetMessageID() string {
+	return cid.fakeCI.ManifestMessageID
+}
+
+func (cid *fakeContainerInstanceDirector) SetLatestSequenceNumber(message *ecsacs.TaskManifestMessage) error {
+	cid.fakeCI.LatestTaskManifestSequenceNumber = aws.Int64Value(message.Timeline)
+	return nil
+}
+
+func (cid *fakeContainerInstanceDirector) SetMessageID(messageID string) error {
+	cid.fakeCI.ManifestMessageID = messageID
+	return nil
+}
+
+// setupTestTaskManifestResponder initializes the TaskManifestResponder for testing.
+func setupTestTaskManifestResponder(t *testing.T, f wsclient.RespondFunc) *taskManifestResponder {
+	// Create task manifest responder.
+	cluster := "test-cluster"
+	ftc := fakeTaskComparer{}
+	cid := fakeContainerInstanceDirector{}
+	cid.fakeCI.LatestTaskManifestSequenceNumber = 15
+	cid.fakeCI.ManifestMessageID = ""
+	testTaskManifestResponder := NewTaskManifestResponder(ftc, &cid, &cid, cluster, metrics.NewNopEntryFactory(), f)
+	return testTaskManifestResponder
+}
+
+// setupTestRespondFunc initializes the respond function used by the TestTaskManifestResponder.
+func setupTestRespondFunc(ackSent *bool, stopVerifSent *bool,
+	stopVerifChannel chan<- *ecsacs.TaskStopVerificationMessage) func(v interface{}) error {
+	f := func(v interface{}) error {
+		_, isAckRequest := v.(*ecsacs.AckRequest)
+		if isAckRequest {
+			*ackSent = true
+		} else {
+			stopVerification, isStopVerification := v.(*ecsacs.TaskStopVerificationMessage)
+			if isStopVerification {
+				*stopVerifSent = true
+				stopVerifChannel <- stopVerification
+			}
+		}
+		return nil
+	}
+	return f
+}
+
+// TestManifestAckHappyPath tests the happy path for a typical TaskManifestMessage and confirms
+// expected ack is made
+func TestManifestAckHappyPath(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ackSent := make(chan *ecsacs.AckRequest)
+
+	testResponseSender := func(response interface{}) error {
+		resp := response.(*ecsacs.AckRequest)
+		ackSent <- resp
+		return nil
+	}
+
+	testTaskManifestResponder := setupTestTaskManifestResponder(t, testResponseSender)
+
+	// Create test task manifest.
+	testManifest := &ecsacs.TaskManifestMessage{
+		ClusterArn:           aws.String(testTaskManifestResponder.cluster),
+		ContainerInstanceArn: aws.String("test-instance-arn"),
+		MessageId:            aws.String("test-messageID"),
+		Tasks: []*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("STOPPED"), TaskArn: aws.String("manifest-task-arn")},
+		},
+		Timeline: aws.Int64(20),
+	}
+
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage) error)
+	go handleManifestMessage(testManifest)
+
+	manifestMessageAckSent := <-ackSent
+	require.Equal(t, aws.StringValue(manifestMessageAckSent.MessageId), aws.StringValue(testManifest.MessageId))
+
```

Comment:
I think we need some more assertions (or in other smaller/focused tests). 

1. TaskStopVerificationMessage is sent or not
2. Contents of the two messages to ACS are correct
3. Sequence number and message ID are updated as expected
4. Metric is published

The latter two should be easy to achieve with mocks generated with gomock. 

Also we should probably add some more test cases to cover non-happy cases.

Reply:
New revision tests 2-4, but #1 will be difficult as the verification is only sent if manifest does not match existing tasks. This can be tested by an implementer that has a data client/dummy data client with access to these instance tasks.

Reply:
Thanks @timj-hh. IIUC `fakeTaskComparer.CompareRunningTasksOnInstanceWithManifest` method's return value decides whether TaskStopVerificationMessage is sent or not. Since we have full control over `fakeTaskComparer`, I think we should be able to write a test case for which TaskStopVerificationMessage is sent?